### PR TITLE
Feature/wave improvement

### PR DESCRIPTION
**Pull request by magnus-gross**
# The Issue

The current wave implementation has a problem, it only is a wave for its background. It does not cut its children UI elements into a circular shape, but into a rectangular shape.
See for example the `SearchBar`:
old
As you can see, only the background works. Every UI component is cut into a rectangular shape.

Another issue is that the underlying implementation caused the children objects to move depending on the size of the wave. This caused the user of a Wave having to manually move the position of children object depending on the wave's size in order to stay at the same position when the wave is animating. You can see this for example in the current `SearchBar` implementation.

This also caused a small performance problem, because it resulted in unnecessary bindings for the x and y property, which were then updated every frame during the animation: http://doc.qt.io/qt-5/qtquick-performance.html#visual-effects

# The fix

The fundamental implementation of Wave had to be changed, otherwise those two issues would not be fixable.
The fix implements the Wave using an `OpacityMask`, which is just a very small OpenGL shader, as you can see here: http://code.qt.io/cgit/qt/qtgraphicaleffects.git/tree/src/effects/OpacityMask.qml#n160

This allows the Wave to cut any children objects in circular shape, which means the first problem is fixed, as you can see here:
new

It simultaneously fixes the second issue, because now children objects are not moved at all in order for the Wave to work.
This means that the user of a Wave can much more comfortably use Waves, since anchors and absolute positions and stuff can be used without having to think about the size of the Wave.

And btw, since I already had the file open, I also documented Wave.qml. :)

### CODE REVIEW

File: src/imports/material/Wave.qml [hunk 1]
```diff
@@ -27,23 +27,54 @@ import QtGraphicalEffects 1.0
 Item {
     id: wave
 
+    /*!
+      \c True, iif the wave is opened
```
Comment (plfiorini) [inline src/imports/material/Wave.qml]: Small typo and should be reformulated like this:

```
Whether wave is open.
```

Also I now realize the property is badly named, it should be open rather opened which is the past of open (http://dictionary.cambridge.org/it/grammatica/grammatica-britannico/open-or-opened)
The property name could be changed with a separate commit.

**Comment by plfiorini** [comment]
Impressive work as usual thanks. I'm going to check out the patches and test it.

**Comment by magnus-gross** [comment]
What about `signal finished(bool opened)`. Should it say `open` there too?

**Comment by magnus-gross** [comment]
The name of `property bool open` collides btw with `function open`

**Comment by magnus-gross** [comment]
I just pushed another commit for renaming `opened`.
Since it collided with the function `open()`, I also renamed the function to `openWave()`.
Tell me, if you had something else in mind

**Comment by plfiorini** [comment]
Why not use `visible` instead of the `open` property?

**Comment by redgrapetech** [comment]
@magnus-gross  
Congratulations and Thanks for improving  fluid.

**Comment by magnus-gross** [comment]
We do not use `visible` in place of `open`, because they are not the same values.
Now, when we talk about opening, there is no difference, but when closing, there is one: We immediately set open to false and visible is only set to false, when the animation finished. Replacing `open` with `visible` would cause the Wave to be immediately invisible, thus making the closing animation invisible as well.
